worker/worker.py

```diff
@@ -215,6 +219,84 @@ def safe_sleep(f):
         print("\nSleep interrupted...")
 
 
+def text_hash(file):
+    with open(file) as f:  # text mode to have newline translation!
+        return base64.b64encode(
+            hashlib.sha384(f.read().encode("utf8")).digest()
+        ).decode("utf8")
+
+
+def generate_sri(install_dir):
+    sri = {}
+    for file in FILE_LIST:
+        item = os.path.join(install_dir, file)
+        sri[file] = text_hash(item)
+    return sri
+
+
+def write_sri(install_dir):
+    sri = generate_sri(install_dir)
+    sri_file = os.path.join(install_dir, "sri.txt")
+    print("Writing sri hashes to {}".format(sri_file))
+    with open(sri_file, "w") as f:
+        json.dump(sri, f)
+        f.write("\n")
+
+
+def verify_sri(install_dir):
+    sri = generate_sri(install_dir)
+    sri_file = os.path.join(install_dir, "sri.txt")
+    if not os.path.exists(sri_file):
+        print("{} does not exist".format(sri_file))
+        return False
+    try:
+        with open(sri_file, "r") as f:
+            sri_ = json.load(f)
+    except Exception as e:
+        print("Exception reading {}:\n".format(sri_file), e, sep="", file=sys.stderr)
+        return False
+    if not isinstance(sri_, dict):
+        print("The file {} does not contain a dictionary".format(sri_file))
+        return False
+    for k, v in sri.items():
+        if k not in sri_ or v != sri_[k]:
+            print("The value for {} is incorrect in {}".format(k, sri_file))
+            return False
+    print("The file {} is up to date!".format(sri_file))
+    return True
+
+
+def download_sri():
+    try:
+        ret = requests_get(SRI_URL).json()
+    except:
+        print("Unable to download {}".format(SRI_URL))
+        return None
+    return ret
+
+
+def verify_remote_sri(install_dir):
+    # Returns:
+    # None in case of network error
+    # True if verification succeeded
+    # False if verification failed
+    sri = generate_sri(install_dir)
+    sri_ = download_sri()
+    if sri_ is None:
+        return None
+    tainted = False
+    for k, v in sri_.items():
```

Comment:
Just for fun :)
```python
if sri_ != sri:
    tainted = True
    for k, v in (set(sri_.items()) - set(sri.items())):
        print("{} has been modified!".format(k))
```

Reply:
Not sure about this one. Personally I find the original code more straightforward and hence more readable...

Reply:
Just as not, the pythonic version is faster in checking the equality but slower in finding the differences:
```python
import timeit
a = {str(i): i for i in range(100)}
#b = {str(i): i+1 for i in range(10)}
b = {str(i): i for i in range(100)}
b["0"]=1

def v():
    for k, v in a.items():
        if k not in b or v != b[k]:
            tainted = True
            #print("{} has been modified!".format(k))

def p():
    if a != b:
        tainted = True
        for k in dict(set(a.items()) ^ set(b.items())):
            pass
            #print("{} has been modified!".format(k)

print("v():", timeit.timeit('[v()]', globals=globals()))
print("p():", timeit.timeit('[p()]', globals=globals()))
```

Reply:
Thanks. Speed is not really a priority for the worker though...

It stands to reason that `dict == dict` is faster than manually going through the key value pairs. In the former case we are executing C code and in the latter case we are executing python code.

Reply:
Of course, I wasted some time to learn&refresh a bit of python. As previously stated, just for fun :)